File: web_version/pmugg_files/network/connectorGroup.js
ms.connectorGroup = function() {
	this.network = null;
	// We actually use the vertexNets attached to the primalVertex which
	// usually represents the connector. It is easier to connect the vertexNets
	// when copying the network.
	this.connectors = [];

	this.id = ms.counter.add('connectorGroup');
};

ms.counter.register('connectorGroup');

ms.connectorGroup.prototype.getNetwork = function() {
	return this.network;
};

ms.connectorGroup.prototype.connectNet = function(network) {
	this.network = network;
	network.addConnectorGroup(this);
	return this;
};

ms.connectorGroup.prototype.addConnector = function(connector, index) {
	this.connectors[index] = connector;
	connector.setGroup(this);
};

ms.connectorGroup.prototype.getConnectors = function() {
	return this.connectors;
};

ms.connectorGroup.prototype.copyConnection = function(copy) {
	var copyNet = copy.getNetwork();
	var self = this;
	this.connectors = copy.getConnectors().map((vertex) => {
		return self.network.convertVertex(copyNet, vertex);
	});
};

ms.connectorGroup.prototype.export = function(copy) {
	var self = this;
	return {
		connectors: this.getConnectors().map((vertex) => {
			return self.network.vertexIndex(vertex);
		}),
	}
};

ms.connectorGroup.prototype.import = function(json) {
	var self = this;
	this.connectors = json.connectors.map((vertex) => {
		return self.network.getVertices()[vertex];
	});
};

File: cpp_version/pmugg/network/connector_group.cpp
#include "connector_group.h"
#include "network.h"
#include "vertex.h"
#include "util.h"

namespace ms {

int ConnectorGroup::nextId = 0;

ConnectorGroup::ConnectorGroup()
    : network(nullptr)
    , id(nextId++) {}

Network* ConnectorGroup::getNetwork() const {
    return network;
}

const std::vector<Vertex*>& ConnectorGroup::getConnectors() const {
    return connectors;
}

int ConnectorGroup::getId() const {
    return id;
}

ConnectorGroup* ConnectorGroup::connectNet(Network* net) {
    network = net;
    network->addConnectorGroup(this);
    return this;
}

void ConnectorGroup::addConnector(Vertex* connector, int index) {
    if (index >= connectors.size()) {
        connectors.resize(index + 1, nullptr);
    }
    connectors[index] = connector;
    if (connector) {
        connector->setGroup(this);
    }
}

void ConnectorGroup::copyConnection(const ConnectorGroup* copy) {
    auto* copyNet = copy->getNetwork();
    auto copyConnectors = copy->getConnectors();
    
    connectors.clear();
    connectors.reserve(copyConnectors.size());
    
    for (auto* vertex : copyConnectors) {
        if (vertex) {
            connectors.push_back(network->convertVertex(copyNet, vertex));
        } else {
            connectors.push_back(nullptr);
        }
    }
}

Json ConnectorGroup::export() const {
    Json json;
    json["connectors"] = Json::array();
    
    for (auto* vertex : connectors) {
        if (vertex) {
            json["connectors"].push_back(network->vertexIndex(vertex));
        } else {
            json["connectors"].push_back(nullptr);
        }
    }
    
    return json;
}

void ConnectorGroup::import(const Json& json) {
    connectors.clear();
    auto& vertices = network->getVertices();
    
    for (const auto& index : json["connectors"]) {
        if (index.is_null()) {
            connectors.push_back(nullptr);
        } else {
            connectors.push_back(vertices[index.get<int>()]);
        }
    }
}

} // namespace ms 